### DISCUSSION TITLE
Expression bodied catch blocks

### DISCUSSION THREAD

**Discussion by ogjtech** [dotnet/csharplang#6402] [Language Ideas]
Sometimes in code, we have multiple `catch`-blocks if we want different exception handling behaviour for different types of exceptions. That behaviour could consist of many lines, but it could also simply be a method call, or an alternative `return` statement.
In many other places in code, i.e. constructors, get-only properties or methods, you could turn it into an expression-bodied function to reduce lines of code and make code generally cleaner.

I think it would be good if this could be done for `catch`-blocks as well.

# Why?
Currently, in my controller layer I am handling specific types of exceptions in actions. 
The code looks like this:
```c-sharp
try
{
    var result = await _manager.Foo(id, ct);

    return result ? NoContent() : BadRequest(ModelState);
}
catch (DataConflictException e)
{
    return Conflict(e);
}
catch (NotFoundException e)
{
    return NotFound(e);
}
catch (BadRequestException e)
{
    return BadRequest(e);
}
catch (UnauthorizedException e)
{
    return Unauthorized(e);
}
catch (Exception e)
{
    return StatusCode(StatusCodes.Status500InternalServerError, e);
}
```

As you can see, the `catch`-blocks currently add 19 extra lines to the function body. I think it would be great if I could write the above codes as below:
```c-sharp
try
{
    var result = await _workOrderTaskManager.SetStatusToFeasible(subsidiaryId, branchId, workOrderId, ct);

    return result ? NoContent() : BadRequest(ModelState);
}
catch (DataConflictException e) => Conflict(e);
catch (NotFoundException e) => NotFound(e);
catch (BadRequestException e) => BadRequest(e);
catch (UnauthorizedException e) => Unauthorized(e);
catch (Exception e) => StatusCode(StatusCodes.Status500InternalServerError, e);
```

Above code reduces the `catch`-blocks to just 5, one line for each block.
Of course, one could also simply create a single exception handling method in which all these different exceptions are handled, but defining `catch`-blocks this way could in general drastically reduce lines of code.

**Comment by KalleOlaviNiemitalo** [comment]
This belongs in <https://github.com/dotnet/csharplang/discussions>.

Catch blocks don't necessarily have to return from the function, so if `catch (NotFoundException e) => NotFound(e);` means `catch (NotFoundException e) { return NotFound(e); }` rather than `catch (NotFoundException e) { NotFound(e); }`, that is somewhat surprising; especially when compared to [switch expression](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/switch-expression), where `return` is not implied.

**Comment by KalleOlaviNiemitalo** [comment]
Similar (but not identical) syntax was proposed in <https://github.com/dotnet/csharplang/discussions/2734>

**Comment by GREsau** [comment]
It's not quite as succinct as your proposal, but you can already achieve something similar with switch expressions:
```csharp
try
{
    var result = await _manager.Foo(id, ct);

    return result ? NoContent() : BadRequest(ModelState);
}
catch (Exception e)
{
    return e switch
    {
        DataConflictException => Conflict(e),
        NotFoundException => NotFound(e),
        BadRequestException => BadRequest(e),
        UnauthorizedException => Unauthorized(e),
        Exception => StatusCode(StatusCodes.Status500InternalServerError, e),
    };
}
```

**Reply by DavidArno** [reply]
To my mind, this is the "correct answer" to this request. No new syntax needed and the solution is very similar to the proposal and follows the same idioms. As a bonus it avoids the problem with the `catch (DataConflictException e) { return Conflict(e); }` approach, which often conflicts with company coding standards that mandate braces on separate lines.

**Reply by HaloFour** [reply]
It works well for the cases where you want to catch and evaluate all types of exceptions, but if you wanted to, say, let `Exception` bubble out, then it would be a bit more complicated.  Unfortunately there's no "rethrow expression", yet, and `ExceptionDispatchInfo.Throw` returns `void` so it can't be mixed into the `switch` expression.

As for workplaces that demand strict formatting in those cases where it makes no sense, I'd suggest advocating for change, but in general C# doesn't add new syntactic forms specifically to work around stylistic preferences.

**Comment by HaloFour** [comment]
This has been legal since C# 1.0:

```cs
try
{
    var result = await _workOrderTaskManager.SetStatusToFeasible(subsidiaryId, branchId, workOrderId, ct);

    return result ? NoContent() : BadRequest(ModelState);
}
catch (DataConflictException e) { return Conflict(e); }
catch (NotFoundException e) { return NotFound(e); }
catch (BadRequestException e) { return BadRequest(e); }
catch (UnauthorizedException e) { return Unauthorized(e); }
catch (Exception e) { return StatusCode(StatusCodes.Status500InternalServerError, e); }
```

The suggested use of `=>` to return a value from the containing method doesn't make much sense.  That's not how `=>` works.  It would imply that the `try` was an expression and that you were returning a value from that expression.
